src/Aspire.Hosting/Dcp/ApplicationExecutor.cs

```diff
@@ -156,21 +158,44 @@ private void WatchResourceChanges(CancellationToken cancellationToken)
 
         async Task WatchKubernetesResource<T>(Func<WatchEventType, T, Task> handler) where T : CustomResource
         {
-            try
+            var retryUntilCancelled = new RetryStrategyOptions()
             {
-                await foreach (var (eventType, resource) in kubernetesService.WatchAsync<T>(cancellationToken: cancellationToken))
+                ShouldHandle = new PredicateBuilder().HandleInner<EndOfStreamException>(),
+                BackoffType = DelayBackoffType.Constant,
+                MaxRetryAttempts = int.MaxValue,
+                UseJitter = true,
+                MaxDelay = TimeSpan.FromSeconds(30),
+                OnRetry = (retry) =>
                 {
-                    await semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+                    _logger.LogDebug(
```

Comment:
This is hard to test. You have to drastically alter KubernetesService to inject this error synthetically and DCP has quite a long timeout for long poll (like Kubernetes). I've asked @karolz-ms for an environment variable we can inject into DCP when launching it to bring down the long poll timeout for resource watching so we can simulate this more easily.